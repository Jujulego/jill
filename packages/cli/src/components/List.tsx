import { Box, Static, Text, useStdout } from 'ink';
import { ReactElement } from 'react';

// Types
export type ListItem<K extends string> = Partial<Record<K, string | number>> & Record<string, unknown>;

export interface ListProps<K extends string> {
	attrs: K[];
	data: ListItem<K>[];
	withoutHeaders?: boolean;
	json?: boolean;
}

// Utils
function capitalize(str: string): string {
	return `${str.at(0)?.toUpperCase()}${str.substring(1)}`;
}

// Component
export const List = <K extends string>(props: ListProps<K>): ReactElement => {
	const { attrs, data, withoutHeaders, json } = props;
	const { stdout } = useStdout();

	// Render
	if (json) {
		const lines = JSON.stringify(data, null, stdout?.isTTY ? 2 : 0).split('\n');

		return (
			<Static items={lines}>
				{ (line, idx) => <Text key={idx}>{ line }</Text> }
			</Static>
		);
	}

	return (
		<Box>
			{attrs.map((attr) => (
				<Box key={attr} flexDirection="column" marginRight={2}>
					{!withoutHeaders && <Text bold>{capitalize(attr)}</Text>}
					{data.map((d, idx) => (
						<Text key={idx} color={d[attr] ? '' : 'grey'}>{d[attr] || 'unset'}</Text>
					))}
				</Box>
			))}
		</Box>
	);
};
