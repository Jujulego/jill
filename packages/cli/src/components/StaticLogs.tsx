import { logger } from '@jujulego/jill-core';
import { FC, useEffect, useState } from 'react';
import { format } from 'winston';
import Transport from 'winston-transport';
import { Static, Text } from 'ink';

// Components
export const StaticLogs: FC = () => {
  // State
  const [logs, setLogs] = useState<any[]>([]);

  // Effects
  useEffect(() => {
    logger.add(new class extends Transport {
      // Constructor
      constructor() {
        super({
          format: format.combine(
            format.errors(),
            format.colorize({ message: true, colors: { debug: 'grey', verbose: 'blue', info: 'white', error: 'red' } }),
          )
        });
      }

      // Methods
      log(log: any, next: () => void): void {
        setLogs((old) => [...old, log]);
        next();
      }
    });
  }, []);

  // Render
  return (
    <Static items={logs}>
      { ({ label, message }, idx) => (
        <Text key={idx}>
          <Text color="grey">jill: { label && `[${label}] ` }</Text>
          { message }
        </Text>
      ) }
    </Static>
  );
};
