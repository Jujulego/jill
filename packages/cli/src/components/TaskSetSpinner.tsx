import { Task, TaskSet } from '@jujulego/jill-core';
import { TaskSpinner } from '@jujulego/jill-common';
import { FC, useEffect, useState } from 'react';

// Types
export interface TaskSetSpinnerProps {
  taskSet: TaskSet;
}

// Component
export const TaskSetSpinner: FC<TaskSetSpinnerProps> = (props) => {
  const { taskSet } = props;

  // State
  const [tasks, setTasks] = useState<Task[]>([]);

  // Effects
  useEffect(() => {
    const handler = (task: Task) => {
      setTasks((old) => [...old, task]);
    };

    taskSet.on('started', handler);

    return () => {
      taskSet.off('started', handler);
    };
  }, [taskSet]);

  // Render
  return (
    <>
      { tasks.map((task, idx) => <TaskSpinner key={idx} task={task} />) }
    </>
  );
};
