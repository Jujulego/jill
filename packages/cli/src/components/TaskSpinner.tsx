import { Task } from '@jujulego/jill-core';
import { Text } from 'ink';
import Spinner from 'ink-spinner';
import logSymbols from 'log-symbols';
import { FC, useEffect, useState } from 'react';

// Types
export interface TaskSpinnerProps {
  task: Task;
}

// Component
export const TaskSpinner: FC<TaskSpinnerProps> = (props) => {
  const { task } = props;

  // State
  const [status, setStatus] = useState(task.status);

  // Effects
  useEffect(() => {
    task.on('status', setStatus);

    return () => {
      task.off('status', setStatus);
    };
  }, [task]);

  // Render
  if (status === 'done') {
    return (
      <Text>
        <Text color="green">{ logSymbols.success }</Text>{' '}
        <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
        { task.name } is { status }
      </Text>
    );
  } else if (status === 'failed') {
    return (
      <Text>
        <Text color="red">{ logSymbols.error }</Text>{' '}
        <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
        { task.name } is { status }
      </Text>
    );
  }

  return (
    <Text>
      <Spinner type={status === 'running' ? 'line' : 'simpleDotsScrolling'} />{' '}
      <Text color="grey">[{ task.context.workspace?.name }]</Text>{' '}
      { task.name } is { status }
    </Text>
  );
};
