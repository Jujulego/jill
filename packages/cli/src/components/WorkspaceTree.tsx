import { Workspace } from '@jujulego/jill-core';
import { Newline, Text, TextProps } from 'ink';
import { FC, ReactElement, useEffect, useState } from 'react';

// Types
export interface WorkspaceTreeProps {
  workspace: Workspace;
  dev?: boolean;
  level?: ReactElement;
}

// Utils
const style = (dev: boolean): TextProps => ({ color: dev ? 'blue' : '' });

// Component
export const WorkspaceTree: FC<WorkspaceTreeProps> = (props) => {
  const { workspace: wks, dev = false, level = '' } = props;

  // State
  const [deps, setDeps] = useState<[Workspace, boolean][]>([]);

  // Effects
  useEffect(() => void (async () => {
    const deps: [Workspace, boolean][] = [];

    for await (const dep of wks.dependencies()) {
      deps.push([dep, dev]);
    }

    for await (const dep of wks.devDependencies()) {
      deps.push([dep, true]);
    }

    setDeps(deps);
  })(), [wks]);

  // Render
  return (
    <Text>
      <Text {...style(dev)}>{ wks.name }</Text>
      { wks.version && (<Text color="grey">@{ wks.version }</Text>) }
      <Newline />

      { deps.map(([dep, isDev], idx) => (
        <Text key={dep.name}>
          { level }<Text {...style(dev)}>{ idx === deps.length - 1 ? '└' : '├'}─{' '}</Text>
          <WorkspaceTree
            workspace={dep}
            dev={isDev}
            level={<>{ level }<Text {...style(dev)}>{ idx === deps.length - 1 ? ' ' : '│' }{'  '}</Text></>}
          />
        </Text>
      )) }
    </Text>
  );
};
