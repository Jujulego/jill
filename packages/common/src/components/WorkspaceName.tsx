import { Workspace } from '@jujulego/jill-core';
import { Text, TextProps } from 'ink';
import { FC } from 'react';

// Types
export interface WorkspaceNameProps extends TextProps {
  workspace: Workspace;
}

// Component
export const WorkspaceName: FC<WorkspaceNameProps> = ({ workspace, ...style }) => (
  <Text {...style}>
    { workspace.name }
    { workspace.version && (<Text color="grey">@{ workspace.version }</Text>) }
  </Text>
);
