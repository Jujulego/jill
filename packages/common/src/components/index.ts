export * from './List';
export * from './TaskSpinner';
export * from './TaskSetSpinner';
