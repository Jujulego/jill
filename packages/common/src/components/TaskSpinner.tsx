import { Task } from '@jujulego/jill-core';
import { Text } from 'ink';
import Spinner from 'ink-spinner';
import logSymbols from 'log-symbols';
import { FC, useEffect, useState } from 'react';
import { WorkspaceName } from './WorkspaceName';

// Types
export interface TaskSpinnerProps {
  task: Task;
}

// Component
export const TaskSpinner: FC<TaskSpinnerProps> = (props) => {
  const { task } = props;
  const wks = task.context.workspace;

  // State
  const [status, setStatus] = useState(task.status);

  // Effects
  useEffect(() => {
    task.on('status', setStatus);

    return () => {
      task.off('status', setStatus);
    };
  }, [task]);

  // Render
  if (status === 'done') {
    return (
      <Text>
        { logSymbols.success }{' '}
        { wks && <><Text color="grey">[<WorkspaceName workspace={wks} />]</Text>{' '}</> }
        { task.name } is { status }
      </Text>
    );
  } else if (status === 'failed') {
    return (
      <Text>
        { logSymbols.error }{' '}
        { wks && <><Text color="grey">[<WorkspaceName workspace={wks} />]</Text>{' '}</> }
        { task.name } is { status }
      </Text>
    );
  }

  return (
    <Text>
      <Spinner type={status === 'running' ? 'line' : 'simpleDotsScrolling'} />{' '}
      { wks && <><Text color="grey">[<WorkspaceName workspace={wks} />]</Text>{' '}</> }
      { task.name } is { status }
    </Text>
  );
};
