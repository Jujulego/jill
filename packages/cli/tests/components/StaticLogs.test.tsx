import { logger } from '@jujulego/jill-core';
import { render } from 'ink-testing-library';

import { StaticLogs } from '../../src/components/StaticLogs';

// Setup
logger.level = 'debug';

// Tests
describe('StaticLogs', () => {
  it('should print logs', async () => {
    expect(logger.transports).toHaveLength(0);

    // First render should add transport
    const { lastFrame, rerender, unmount } = render(<StaticLogs />);
    await new Promise(res => setTimeout(res, 0));

    expect(logger.transports).toHaveLength(1);

    // Should print logs
    logger.debug('Test debug log');
    logger.verbose('Test verbose log');
    logger.info('Test info log');
    logger.warn('Test warn log');
    logger.error('Test error log');

    rerender(<StaticLogs />);

    expect(lastFrame()).toMatchSnapshot();

    // Unmount should remove transport
    unmount();
    await new Promise(res => setTimeout(res, 0));

    expect(logger.transports).toHaveLength(0);
  });

  it('should print logs with labels', async () => {
    const { lastFrame, rerender, unmount } = render(<StaticLogs />);
    await new Promise(res => setTimeout(res, 0));

    // Should print logs
    logger.debug('Test debug log', { label: 'test' });
    logger.verbose('Test verbose log', { label: 'test' });
    logger.info('Test info log', { label: 'test' });
    logger.warn('Test warn log', { label: 'test' });
    logger.error('Test error log', { label: 'test' });

    rerender(<StaticLogs />);

    expect(lastFrame()).toMatchSnapshot();

    // Unmount should remove transport
    unmount();
    await new Promise(res => setTimeout(res, 0));
  });
});
