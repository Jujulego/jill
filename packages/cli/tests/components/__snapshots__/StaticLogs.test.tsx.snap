// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticLogs should print logs 1`] = `
"[90mjill: [39m[90mTest debug log[39m
[90mjill: [39m[34mTest verbose log[39m
[90mjill: [39m[37mTest info log[39m
[90mjill: [39m[33mTest warn log[39m
[90mjill: [39m[31mTest error log[39m
"
`;

exports[`StaticLogs should print logs with labels 1`] = `
"[90mjill: [test] [39m[90mTest debug log[39m
[90mjill: [test] [39m[34mTest verbose log[39m
[90mjill: [test] [39m[37mTest info log[39m
[90mjill: [test] [39m[33mTest warn log[39m
[90mjill: [test] [39m[31mTest error log[39m
"
`;
